sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ConcurrentConsumer.java

```diff
@@ -0,0 +1,231 @@
+    this.consumer = consumer;
+    this.pollDuration = pollDuration;
+    this.times = AtomicLongMap.create();
+    this.positions = AtomicLongMap.create();
+    this.recordsLagMax =
+        Suppliers.memoize(
+            () ->
+                this.consumer.metrics().values().stream()
+                    .filter(
+                        m ->
+                            "consumer-fetch-manager-metrics".equals(m.metricName().group())
+                                && "records-lag-max".equals(m.metricName().name())
+                                && !m.metricName().tags().containsKey("topic")
+                                && !m.metricName().tags().containsKey("partition"))
+                    .findAny()
+                    .get());
+    this.partitionRecordsLag = new ConcurrentHashMap<>();
+    this.pollResult = ConsumerRecords.empty();
+    this.offsetsForTimesResult = Collections.emptyMap();
+  }
+
+  @Override
+  public void close() {
+    this.phaser.forceTermination();
+    try {
+      this.consumer.wakeup();
+      this.consumer.close();
+    } catch (Exception e) {
+      LOG.error("Exception while closing Kafka consumer", e);
+    }
+    this.times.clear();
+    this.positions.clear();
+    this.pollResult = ConsumerRecords.empty();
+    this.offsetsForTimesResult = Collections.emptyMap();
+  }
+
+  boolean isClosed() {
+    return this.phaser.isTerminated();
+  }
+
+  Map<MetricName, ? extends Metric> metrics() {
+    return this.consumer.metrics();
+  }
+
+  long currentLagOrMaxLag(final TopicPartition topicPartition) {
+    final Supplier<Metric> metric =
+        this.partitionRecordsLag.getOrDefault(topicPartition, this.recordsLagMax);
+    try {
+      return ((Number) metric.get().metricValue()).longValue();
+    } catch (Exception e) {
+      return 0;
+    }
+  }
+
+  long position(final TopicPartition topicPartition) {
+    return this.positions.get(topicPartition) & Long.MAX_VALUE;
+  }
+
+  long initialOffsetForPartition(final TopicPartition topicPartition) {
+    // Offsets start at 0 and there is no position to advance to beyond Long.MAX_VALUE.
+    // The sign bit indicates that an assignment is paused.
+    checkState(this.phaser.register() >= 0);
+    this.positions.put(topicPartition, Long.MIN_VALUE);
+
+    // Synchronize and throw if the consumer was terminated in between.
+    checkState(this.phaser.arriveAndAwaitAdvance() >= 0);
+
+    // Removal will revoke the assignment when the phase advances.
+    final long result = this.positions.remove(topicPartition);
+
+    // Synchronize and ignore the consumer status since the result is already known.
+    this.phaser.arriveAndDeregister();
+
+    // Since Long.MIN_VALUE only has the sign bit set, this will return 0 as a default value if no
+    // position could be determined.
+    return result & Long.MAX_VALUE;
+  }
+
+  @Nullable
+  OffsetAndTimestamp initialOffsetForTime(final TopicPartition topicPartition, final long time) {
```

Comment:
what is units for time? can Instant be used here and changed to long internally?

Reply:
It's in milliseconds, the types requested by the `ConcurrentConsumer` wrapper match the client library where possible: https://kafka.apache.org/38/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#offsetsForTimes(java.util.Map)